crates/runtime/src/lib.rs

```diff
@@ -356,8 +357,45 @@ impl_runtime_apis! {
             Some((tx_type, events))
         }
 
+        fn get_tx_messages_to_l1(tx_hash: TransactionHash) -> Vec<MessageToL1> {
+            Starknet::tx_messages(tx_hash)
+        }
+
         fn get_tx_execution_outcome(tx_hash: TransactionHash) -> Option<Vec<u8>> {
-           Starknet::tx_revert_error(tx_hash).map(|s| s.into_bytes())
+            Starknet::tx_revert_error(tx_hash).map(|s| s.into_bytes())
+        }
+
+        fn get_starknet_messages(block_extrinsics: Vec<<Block as BlockT>::Extrinsic>, chain_id: Felt252Wrapper) -> (Vec<MessageL2ToL1>, Vec<MessageL1ToL2>) {
+            let mut messages_to_l1: Vec<MessageL2ToL1> = Vec::new();
+            let mut messages_to_l2: Vec<MessageL1ToL2> = Vec::new();
+
+            for xt in block_extrinsics {
+                let tx_hash = match xt.function {
+                    RuntimeCall::Starknet( invoke { transaction }) => {
+                        Some(transaction.compute_hash::<<Runtime as Config>::SystemHash>(chain_id, false))
+                    },
+                    RuntimeCall::Starknet( declare { transaction, .. }) => {
+                        Some(transaction.compute_hash::<<Runtime as Config>::SystemHash>(chain_id, false))
+                    },
+                    RuntimeCall::Starknet( deploy_account { transaction }) => {
+                        Some(transaction.compute_hash::<<Runtime as Config>::SystemHash>(chain_id, false))
+                    },
+                    RuntimeCall::Starknet( consume_l1_message { transaction, .. }) => {
+                        let tx_hash = Some(transaction.compute_hash::<<Runtime as Config>::SystemHash>(chain_id, false));
+                        messages_to_l2.push(transaction.into());
+                        tx_hash
+                    },
+                    _ => None
+                };
+                if let Some(tx_hash) = tx_hash {
```

Comment:
I don't think declare/deploy transactions can have messages tho

Reply:
Doesn't it depend on the account contract used? 
Most probably don't, but is it really impossible?

Reply:
I though abt it, technically there should be no limitations? Same goes to the fee execution

Reply:
The fee execution is done through a hardcoded contract configured by the appchain maintainers. So we can know for sure it does not produce a message.